### PR TITLE
Add flaky_tests.py and flaky-tests.yaml

### PR DESCRIPTION
**Pull request by smatting**
This PR adds a a script `hack/bin/flaky_test.py` that lists flaky integration tests together with links to the builds they failed. It lists flaky tests that caused the most builds to fail first.

The script makes use of the data collected by the new [failing-tests](https://concourse.ops.zinfra.io/teams/main/pipelines/failing-tests) pipeline. Since this pipeline only just started collecting data the quality of reporting from `flaky_test.py` should increase when more data has been collected.

Flaky tests are discovered from data and also can manually marked as flaky by adding entries to `flaky-tests.yaml`. You can also add comments to `flaky-tests.yaml` that are shown in the output of the script. Flaky test discovery works by assuming that all failing test in the `staging`, `prod` and `mls` pipelines are flaky.

The script tries to fetch data from the last 4 months. We could improve the script by making this configurable and add caching (the data collected by `failing-tests` is stored per-week).

Example output:

```
❄ "team tests around truncation limits - no events, too large team"
    Exception: Timeout: No matching notification received.
    Match failure: expected: Just "user.delete"
     but got: Just "conversation.delete"

  · 2023-03-23 (today) https://concourse.ops.zinfra.io/teams/main/pipelines/wire-server-pr/jobs/integration-test>
  · 2023-03-22 (yesterday) https://concourse.ops.zinfra.io/teams/main/pipelines/mls/jobs/test/builds/57.4
  · 2023-03-21 (2 days ago) https://concourse.ops.zinfra.io/teams/main/pipelines/wire-server-pr/jobs/integration>
  · 2023-03-21 (2 days ago) https://concourse.ops.zinfra.io/teams/main/pipelines/wire-server-pr/jobs/integration>

❄ "POST /register - can add team members above fanout limit when whitelisting is enabled"
    (discovered flake, no notes in flaky-tests.yaml)

  · 2023-03-22 (yesterday) https://concourse.ops.zinfra.io/teams/main/pipelines/mls/jobs/test/builds/57.4
  · 2023-03-22 (yesterday) https://concourse.ops.zinfra.io/teams/main/pipelines/wire-server-pr/jobs/integration->
  · 2023-03-21 (2 days ago) https://concourse.ops.zinfra.io/teams/main/pipelines/wire-server-pr/jobs/integration>

❄ "GET /mls/key-packages/claim/local/:user - self claim"
    (discovered flake, no notes in flaky-tests.yaml)

  · 2023-03-22 (yesterday) https://concourse.ops.zinfra.io/teams/main/pipelines/mls/jobs/test/builds/57.2
  · 2023-03-21 (2 days ago) https://concourse.ops.zinfra.io/teams/main/pipelines/mls/jobs/test/builds/57
  · 2023-03-21 (2 days ago) https://concourse.ops.zinfra.io/teams/main/pipelines/wire-server-pr/jobs/integration>

❄ "POST /federation/on-user-deleted-conversations : Remove deleted remote user from local conversations"
    (discovered flake, no notes in flaky-tests.yaml)

  · 2023-03-22 (yesterday) https://concourse.ops.zinfra.io/teams/main/pipelines/wire-server-pr/jobs/integration->
  · 2023-03-22 (yesterday) https://concourse.ops.zinfra.io/teams/main/pipelines/mls/jobs/test/builds/57.3
...
```

### CODE REVIEW

File: hack/bin/flaky_tests.py [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+        if n == 0:
+            return 'today'
+        elif n == 1:
+            return 'yesterday'
+        else:
+            return f'{n} days ago'
+    else:
+        weeks = n // 7
+        if weeks < 4:
+            return f'{weeks} weeks ago'
```
Comment (pcapriotti) [inline hack/bin/flaky_tests.py:127]:
```suggestion
            return f'{weeks} week{'s' if weeks >= 2 else ''} ago'
```

File: hack/bin/flaky_tests.py [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+        else:
+            return f'{weeks // 4} months ago'
```
Comment (pcapriotti) [inline hack/bin/flaky_tests.py:129]:
```suggestion
            return f'{weeks // 4} month{'s' if weeks >= 8 else ''} ago'
```

File: hack/bin/flaky_tests.py [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+        s = human_format_date(end_time, today)
+        if (today - end_time) < datetime.timedelta(days=CONCOURSE_LOG_RETENTION_DAYS):
+            url = create_url(b)
+            s = s + ' ' + url
+        lines.append('  ' + s)
+
+    return "\n".join(lines) + '\n'
+
+def pager(s):
+    pipe = os.popen('less -S', 'w')
```
Comment (pcapriotti) [inline hack/bin/flaky_tests.py:160]:
```suggestion
    pipe = os.popen('less -RS', 'w')
```
Adding `-R` here makes `less` render escape sequences correctly.

**Review by pcapriotti — APPROVED** [review]
This is great! What seems to be missing to this setup (or maybe I'm misunderstanding how this is meant to be used) is a way to mark a test as **not** a flake, so we don't see it in the output of the tool.

Few nits below.

**Comment by supersven** [comment]
I'm wondering if we should use type annotations in Python code: https://docs.python.org/3/library/typing.html ? :thinking:

The only drawbacks I see are:
- More typing on the keyboard :wink: 
- Won't work with old Python versions

The latter issue is probably solved by our nix env where we can easily decide which version to take.

**Comment by smatting** [comment]
> What seems to be missing to this setup (or maybe I'm misunderstanding how this is meant to be used) is a way to mark a test as **not** a flake, so we don't see it in the output of the tool.

This is necessary, because sometimes we (I) have merged PR with broken tests. We can add it in a followup PR.

I think it's more the other way around: not all flaky tests are discovered, because they haven't yet occured in staging/mls/prod but on wire-server-pr. Another followup could be to list all the failing tests that are not marked or discovered to be flaky.
But I hope we will have the convention to add flaky tests to `flaky-tests.yaml` as soon as we discover them (or even fix them right away).

**Pull request merged**